driver-core/src/main/com/mongodb/internal/operation/QueryBatchCursor.java

```diff
@@ -263,49 +262,47 @@ public int getMaxWireVersion() {
     }
 
     private void getMore() {
-        Connection connection = getConnection();
+        ServerCursor serverCursor = assertNotNull(resourceManager.serverCursor());
+        Connection connection = resourceManager.connection();
         try {
+            ServerCursor nextServerCursor;
             if (serverIsAtLeastVersionThreeDotTwo(connection.getDescription())) {
                 try {
-                    initFromCommandResult(connection.command(namespace.getDatabaseName(),
-                            asGetMoreCommandDocument(),
+                    nextServerCursor = initFromCommandResult(connection.command(namespace.getDatabaseName(),
+                            asGetMoreCommandDocument(serverCursor),
                             NO_OP_FIELD_NAME_VALIDATOR,
                             ReadPreference.primary(),
                             CommandResultDocumentCodec.create(decoder, "nextBatch"),
-                            connectionSource.getSessionContext(),
-                            connectionSource.getServerApi()));
+                            resourceManager.sessionContext(),
+                            resourceManager.serverApi()));
                 } catch (MongoCommandException e) {
                     throw translateCommandException(e, serverCursor);
                 }
             } else {
                 QueryResult<T> getMore = connection.getMore(namespace, serverCursor.getId(),
                         getNumberToReturn(limit, batchSize, count), decoder);
-                initFromQueryResult(getMore);
+                nextServerCursor = initFromQueryResult(getMore);
             }
+            resourceManager.setServerCursor(nextServerCursor);
             if (limitReached()) {
-                killCursor(connection);
+                resourceManager.releaseServerAndClientResources(connection);
             }
         } catch (MongoSocketException e) {
-            // If connection is pinned, don't attempt to kill the cursor on close, since the connection is in a bad state
-            if (this.connection != null) {
-                killCursorOnClose = false;
+            try {
+                resourceManager.onCorruptedConnection(connection);
+            } catch (RuntimeException suppressed) {
+                e.addSuppressed(suppressed);
             }
             throw e;
         } finally {
             connection.release();
```

Comment:
This seems like it should be the responsibility of ResourceManager to release the Connection, since it's the one that has retained it.

Reply:
When we use the method `ResourceManager.connection`, the `ResourceManager` acts as a `ConnectionSource`, and just like with the `ConnectionSource`, the object that provides a connection retains it, and caller must release it<sup>*</sup>. However, this indeed can be changed if a method `ResourceManager.doWithConnection` is introduced. Will do.

----
<sup>*</sup> This is a consequence of the `ReferenceCounted` contract, which states that a newly created object has count 1 and must be released even if the user of the object did not retained the object explicitly. I do not think there is another way `ReferenceCounted` can be specified.

Reply:
Done. I think the code does look simpler now.

Reply:
I like it!